src/linker/Linker/ConsoleLogger.cs

```diff
@@ -2,15 +2,59 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
+using System.Collections.Generic;
+using System.IO;
+using Mono.Cecil;
 
 namespace Mono.Linker
 {
 	public class ConsoleLogger : ILogger
 	{
-		public void LogMessage (MessageContainer message)
+		public Dictionary<MemberReference, string> ComputedStrings;
+
+		readonly List<MessageContainer> _messageContainers;
+		readonly MessageCategory _categoriesToCache;
+		readonly StreamWriter _streamWriter;
+
+		public ConsoleLogger (StreamWriter streamWriter, MessageCategory categoriesToCache = MessageCategory.None)
+		{
+			ComputedStrings = new Dictionary<MemberReference, string> ();
+			_messageContainers = new List<MessageContainer> ();
+			_categoriesToCache = categoriesToCache;
+			_streamWriter = streamWriter;
+		}
+
+		public void LogMessage (MessageContainer messageContainer)
+		{
+			if (_categoriesToCache.HasFlag (messageContainer.Category)) {
+				_messageContainers.Add (messageContainer);
+				return;
+			}
+
+			_streamWriter?.WriteLine (messageContainer.ToString ());
+		}
+
+		public void Flush ()
+		{
+			if (_streamWriter is null)
```

Comment:
When will this happen and why don't check this earlier?

Reply:
With the changes I just pushed there's no way this can happen. I originally put this so that we could swallow all the output in the test infra (I'm now redirecting `Console` stdout to a `StringWriter`).